janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexFilterOptimizerStrategy.java

```diff
@@ -32,73 +32,179 @@
  */
-public class AdjacentVertexFilterOptimizerStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy> implements TraversalStrategy.ProviderOptimizationStrategy {
+public class AdjacentVertexFilterOptimizerStrategy
+    extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy>
+    implements TraversalStrategy.ProviderOptimizationStrategy {
 
-    private static final AdjacentVertexFilterOptimizerStrategy INSTANCE = new AdjacentVertexFilterOptimizerStrategy();
+    private static final AdjacentVertexFilterOptimizerStrategy INSTANCE =
+        new AdjacentVertexFilterOptimizerStrategy();
 
-    private AdjacentVertexFilterOptimizerStrategy() {
-    }
+    private AdjacentVertexFilterOptimizerStrategy() {}
 
-    public static AdjacentVertexFilterOptimizerStrategy instance() {
-        return INSTANCE;
-    }
+    public static AdjacentVertexFilterOptimizerStrategy instance() { return INSTANCE; }
 
+    private enum OptimizableQueryType { NONE, IS, HASID }
+    ;
 
     @Override
     public void apply(final Traversal.Admin<?, ?> traversal) {
 
-        TraversalHelper.getStepsOfClass(TraversalFilterStep.class, traversal).forEach(originalStep -> {
-            // Check if this filter traversal matches the pattern: _.inV/outV/otherV.is(x)
-            Traversal.Admin<?, ?> filterTraversal = (Traversal.Admin<?, ?>) originalStep.getLocalChildren().get(0);
-            List<Step> steps = filterTraversal.getSteps();
-            if (steps.size() == 2 &&
-                    (steps.get(0) instanceof EdgeVertexStep || steps.get(0) instanceof EdgeOtherVertexStep) &&
-                    (steps.get(1) instanceof IsStep)) {
-                //Get the direction in which we filter on the adjacent vertex (or null if not a valid adjacency filter)
-                Direction direction = null;
-                if (steps.get(0) instanceof EdgeVertexStep) {
-                    EdgeVertexStep evs = (EdgeVertexStep) steps.get(0);
-                    if (evs.getDirection() != Direction.BOTH) direction = evs.getDirection();
-                } else {
-                    assert steps.get(0) instanceof EdgeOtherVertexStep;
-                    direction = Direction.BOTH;
-                }
-                P predicate = ((IsStep) steps.get(1)).getPredicate();
-                //Check that we have a valid direction and a valid vertex filter predicate
-                if (direction != null && predicate.getBiPredicate() == Compare.eq && predicate.getValue() instanceof Vertex) {
-                    Vertex vertex = (Vertex) predicate.getValue();
-
-                    //Now, check that this step is preceded by VertexStep that returns edges
-                    Step<?, ?> currentStep = originalStep.getPreviousStep();
-                    while (currentStep != EmptyStep.instance()) {
-                        if (!(currentStep instanceof HasStep) && !(currentStep instanceof IdentityStep)) {
```

Comment:
I removed the repeated jumps because issues arose when using this strategy on queries that utilize `SideEffectStep`s.
The re-ordering of steps caused the logical order of these steps to break.

Reply:
As I understand you found a bug which currently exists. Would you be able to provide a separate PR which removes this jumps and adds a test where we can see that bug? Or is the appears only in optimized queries with .hasId(id)?

Reply:
If the bug can be reproduced only in this PR, would you be able to add a test to this PR which shows the issue you described?

Reply:
This is a bug I encountered when applying the optimization techniques from this class in PR #1961 
As the basic idea behind both approaches is essentially the same, it should be possible to find a matching test case. I'll try to find one once I get to work on this the next time.

Reply:
Nevermind, I introduced the bug myself when I was trying to catch even more optimizable sequences of steps. Thanks for your review!